ci: update lint workflow

rokm opened the pull request:
I've noticed our `lint` workflow is more lenient than my local runs of `flake8`. Turns out this is because it is running python 3.8 and is not using the latest version of `flake8` (which dropped support for python 3.8 in v7.2.0).

Also clean up the workflow .yml files a bit by removing the `v4` branch from `on.*.branches` list.

rokm committed:
ci: remove 'v4' branch from list of branches that trigger workflows

rokm committed:
ci: update lint workflow

Update `lint` workflow to use `python-version: 3.x` with
`actions/setup-python`, in order to install the latest available
python version insteadf of 3.8.

This ensures that we get up-to-date version of tools (for example,
`flake8` >= 7.2.0 requires python >= 3.9).

Bump the pinned version of `yapf` from 0.32.0 to 0.43.0.

rokm committed:
fake-modules/pyi_splash: remove redundant global variable

The global variable `_ipc_socket` in `_initialize()` is never
assigned, so it does not need to be declared as global variable
at all.

Contemporary versions of `flake8` give a warning about this:

```
F824 `global _ipc_socket` is unused: name is never assigned in scope
    global _initialized, _ipc_socket, _ipc_socket_close
```

rokm committed:
tests: do not install pytest-rerunfailures v16.0

See: https://github.com/pytest-dev/pytest-rerunfailures/issues/302

bwoodsend commented on .github/workflows/ci.yml:
```diff
@@ -6,7 +6,6 @@ on:
   push:
     branches:
       - 'develop'
-      - 'v4'
```
I know I asked this before but can't we just get rid of the `push` trigger?

rokm replied:
You mean so that we don't re-run the tests after merging a PR (which under normal circumstances would undergo a test run before the merge due to `pull_request` trigger)?

bwoodsend replied:
Yeah, it just clogs up CI whenever multiple changes are being proposed (including in the hooks repo).

bwoodsend replied:
Then often the *cancel previous jobs* job can't run because it's stuck in the queue too.

rokm replied:
Let's try removing it, then. 

I'll add a commit here (I now see I should modify the one that removed v4 references from `pull_request` to also remove references from the comments above..).

The main benefit of the "post-merge" test run as I see it is to ensure that the tip of the `develop` branch is always tested. But since all modifications go through PRs, the testing on PRs should guarantee this as well, at least in most cases (I am cynical enough to imagine a situation where we could end up merging multiple PRs that pass the tests on their own, but somehow manage to end up interfering with each other...)

rokm committed:
ci: remove the "push" trigger for the main CI workflow

Run the main CI workflow only on pull requests against the
`develop` branch, i.e., avoid re-running it on the `develop`
branch itself after a merge.
